rustls/src/crypto/mod.rs

```diff
@@ -505,6 +505,81 @@ pub trait ActiveKeyExchange: Send + Sync {
+    /// using that key share.  If not, the server sends a `HelloRetryRequest` instructing
+    /// the client to send a different key share instead.
+    ///
+    /// This costs an extra round trip, and wastes the key exchange computation
+    /// (in [`SupportedKxGroup::start()`]) the client already did.  We would
+    /// like to avoid those wastes if possible.
+    ///
+    /// It is early days for post-quantum-secure hybrid key exchange deployment.
+    /// This means (commonly) continuing to offer both the hybrid and classical
+    /// key exchanges, so the handshake can be completed without a `HelloRetryRequest`
+    /// for servers that support the offered hybrid or classical schemes.
+    ///
+    /// Implementing `hybrid_pair` enables two optimizations:
+    ///
+    /// 1. Sending both the hybrid and classical key shares in the `ClientHello`.
+    ///
+    /// 2. Performing the classical key exchange setup only once.  This is important
+    ///    because the classical key exchange setup is relatively expensive.
+    ///    This optimization is permitted and described in
+    ///    <https://www.ietf.org/archive/id/draft-ietf-tls-hybrid-design-10.html#section-3.2>
+    ///
+    /// Both of these only happen if the classical algorithm appears separately in
+    /// the client's [`CryptoProvider::kx_groups`], and if the hybrid algorithm appears
+    /// first in that list.
+    ///
+    /// # How it works
+    /// This function is only called by rustls for clients.  It is called when
+    /// constructing the initial `ClientHello`.  rustls follows these steps:
+    ///
+    /// 1. If the return value is None, nothing further happens.
+    /// 2. If the given [`ActiveKeyExchange::group()`] does not appear in
+    ///    [`CryptoProvider::kx_groups`], nothing further happens.
+    /// 3. The given [`ActiveKeyExchange::pub_key()`] is added to the `ClientHello`,
+    ///    after the hybrid entry.
+    ///
+    /// Then, one of three things may happen when the server replies to the `ClientHello`:
+    ///
+    /// 1. The server sends a `HelloRetryRequest`.  Everything is thrown away and
+    ///    we start again.  Note that `hybrid_pair` is never called in this case.
+    /// 2. The server agrees to our hybrid key exchange: rustls calls
+    ///    [`ActiveKeyExchange::complete()`] consuming `self`.  `into_hybrid_pair`
+    ///    is not called.
+    /// 3. The server agrees to our classical key exchange: rustls calls
+    ///    [`ActiveKeyExchange::into_hybrid_pair()`] which discards the hybrid
+    ///    key data, and yields just the classical key exchange which is then
+    ///    consumed by [`ActiveKeyExchange::complete()`].
+    fn hybrid_pair(&self) -> Option<&dyn ActiveKeyExchange> {
+        None
+    }
+
+    /// Consumes this hybrid key exchange, yielding the underlying classical
+    /// key exchange.
+    ///
+    /// If [`ActiveKeyExchange::hybrid_pair()`] returns non-None, this
+    /// _must_ return non-None.
```

Comment:
nit: I would find it easier to parse this without "non-None", maybe  "returns Some(x), this _must_ return Some(x)"`?

Should there also be some mention that it should be returning the _same_ `dyn ActiveKeyExchange` as `hybrid_pair()`?